go2ts.go

```diff
@@ -96,38 +128,90 @@ func (g *Go2TS) tSTypeFromStructFieldType(reflectType reflect.Type) *tsType {
 		kind = reflectType.Kind()
 	}
 
-	// Default to setting the tsType from the Go type.
-	ret.typeName = reflectTypeToTypeScriptType(reflectType)
+	// As we build up the chain of tsTypes that fully describes a typeDefinition
+	// we may come across named types. For example: map[string]Donut, where
+	// Donut could be a "type Donut struct {...}", or a type based on a
+	// primitive type, such as "type Donut string". In this case we need to add
+	// that type to all of our known types and return a reference to that type
+	// from here.
+	if !calledFromAddType && // Don't do this if called from addType().
+		reflectType.Name() != "" && // Don't bother with anonymous structs.
+		!isTime(reflectType) && // Also skip time.Time.
+		(!isPrimitive(reflectType.Kind()) || // And either it's not a primitive Kind.
+			// Or it's case where a primitive Kind like string shows up with a type name.
+			(isPrimitive(reflectType.Kind()) && reflectType.Name() != reflectType.Kind().String())) {
+		typeName, err := g.addType(reflectType, "")
+		if err == nil {
+			return &tsType{
+				typeName:  typeName,
+				canBeNull: ret.canBeNull,
+			}
+		}
+	}
+
+	// Default to using the Go lang type name as the TypeScript type name.
+	nativeType := reflectType.String()
+
+	// Strip off the module name of the native type if present.
+	if i := strings.IndexByte(nativeType, '.'); i > -1 {
+		nativeType = nativeType[i+1:]
+	}
+	ret.typeName = nativeType
 
-	// Update the type if the kind points to something besides a primitive type.
 	switch kind {
+	case reflect.Uint8,
+		reflect.Uint16,
+		reflect.Uint32,
+		reflect.Uint64,
+		reflect.Uint,
+		reflect.Int8,
+		reflect.Int16,
+		reflect.Int32,
+		reflect.Int64,
+		reflect.Int,
+		reflect.Float32,
+		reflect.Float64:
+		ret.typeName = "number"
+
+	case reflect.String:
+		ret.typeName = "string"
+
+	case reflect.Bool:
+		ret.typeName = "boolean"
+
 	case reflect.Map:
 		ret.typeName = "map"
-		ret.keyType = reflectTypeToTypeScriptType(reflectType.Key())
-
-		ret.subType = g.tSTypeFromStructFieldType(reflectType.Elem())
+		ret.keyType = g.tsTypeFromReflectType(reflectType.Key(), false)
+		ret.subType = g.tsTypeFromReflectType(reflectType.Elem(), false)
 
 	case reflect.Slice, reflect.Array:
 		ret.typeName = "array"
 		ret.canBeNull = (kind == reflect.Slice)
-
-		ret.subType = g.tSTypeFromStructFieldType(reflectType.Elem())
+		ret.subType = g.tsTypeFromReflectType(reflectType.Elem(), false)
 
 	case reflect.Struct:
 		if isTime(reflectType) {
-			break
+			ret.typeName = "string"
+		} else {
+			ret.typeName = "interface"
+			name, _ := g.addType(reflectType, "")
+			ret.interfaceName = name
 		}
-		ret.typeName = "interface"
-		st, _ := g.addType(reflectType, "")
-		ret.interfaceName = st.Name
 
 	case reflect.Interface:
 		ret.typeName = "any"
+
+	case reflect.Complex64,
+		reflect.Complex128,
+		reflect.Chan,
+		reflect.Func,
+		reflect.UnsafePointer:
+		panic(fmt.Sprintf("Go kind %q can't be serialized to JSON.", kind))
```

Comment:
Nit: `"Go Kind %q cannot be converted to TypeScript."`.

Reply:
Done.